text/0021-accessibility.md

```diff
@@ -0,0 +1,485 @@
+
+    use iced_accessibility::accesskit::{
+        Node, NodeId, Role, Tree, TreeUpdate,
+    };
+    use iced_accessibility::accesskit_winit::Adapter;
+    let title = state.title().to_string();
+    let proxy_clone = proxy.clone();
+    (
+        node_id,
+        Adapter::new(
+            &window,
+            move || {
+                let _ =
+                    proxy_clone.send_event(UserEventWrapper::A11yEnabled);
+                TreeUpdate {
+                    nodes: vec![(
+                        NodeId(node_id),
+                        std::sync::Arc::new(Node {
+                            role: Role::Window,
+                            name: Some(title.into_boxed_str()),
+                            ..Default::default()
+                        }),
+                    )],
+                    tree: Some(Tree::new(NodeId(node_id))),
+                    focus: None,
+                }
+            },
+            proxy.clone(),
+        ),
+    )
+};
+```
+- adding a boolean that tracks whether accessibility has been enabled
+- sending accessibility tree updates to accesskit if accessibility has been enabled
+```rust
+#[cfg(feature = "a11y")]
+if a11y_enabled {
+    use iced_accessibility::{
+        accesskit::{Node, NodeId, Role, Tree, TreeUpdate},
+        A11yId, A11yNode, A11yTree,
+    };
+    // TODO send a11y tree
+    let child_tree = user_interface.a11y_nodes();
+    let root = Node {
+        role: Role::Window,
+        name: Some(state.title().to_string().into_boxed_str()),
+        ..Default::default()
+    };
+    let window_tree = A11yTree::node_with_child_tree(
+        A11yNode::new(root, window_a11y_id),
+        child_tree,
+    );
+    let tree = Tree::new(NodeId(window_a11y_id));
+    let mut current_operation =
+        Some(Box::new(OperationWrapper::Id(Box::new(
+            operation::focusable::find_focused(),
+        ))));
+    let mut focus = None;
+    while let Some(mut operation) = current_operation.take() {
+        user_interface.operate(&renderer, operation.as_mut());
+
+        match operation.finish() {
+            operation::Outcome::None => {}
+            operation::Outcome::Some(message) => match message {
+                operation::OperationOutputWrapper::Message(
+                    _,
+                ) => {
+                    unimplemented!();
+                }
+                operation::OperationOutputWrapper::Id(id) => {
+                    focus = Some(A11yId::from(id));
+                }
+            },
+            operation::Outcome::Chain(next) => {
+                current_operation = Some(Box::new(
+                    OperationWrapper::Wrapper(next),
+                ));
+            }
+        }
+    }
+
+    adapter.update(TreeUpdate {
+        nodes: window_tree.into(),
+        tree: Some(tree),
+        focus: focus.map(|id| id.into()),
+    });
+}
+```
+- reset widget Id counter before building the UserInterface
```

Comment:
This seems very brittle to me as the IDs are tied to the order in which widgets are created. If for some reason a widget change its place in the "creation chain" during one frame, AccessKit will be confused.

Reply:
From what I recall, egui also computes a hash of some sort to generate IDs, which can be valid as well even though there's a tiny chance of collision.
If you can retain IDs between frames, it's indeed safer.